### ISSUE TITLE
Adding models without recompiling infer

### ISSUE THREAD

**Issue by sumerman** [facebook/infer#456]
Is it even possible to add a new model without recompiling? From what I can tell there is special undocumented `models-mode` and a global specs folder. Is that right? How hard would it be to allow to specify multiple spec folders? Is there a good place to start digging?

**Comment by akotulski** [comment]
Currently it's not possible, but it shouldn't be very hard to add an option to specify extra search path. Right now path is hardcoded [here](https://github.com/facebook/infer/blob/13de8a0835ac3cf04c92dc73b2602e6aa0ece29e/infer/src/backend/config.ml#L242).

The other part is to create those extra specs, but I think running infer with `--models-mode` may be enough.

Could you provide more context on what you are trying to do? Each language(C/C++/Java) has differences in modeling mechanisms.

**Comment by sumerman** [comment]
Ultimately I want to apply infer to check things like PostgreSQL extensions, but, for now, I'm just playing around with `libpq` and a small program using it. 

Currently, I'm struggling with making infer detect simple resource leak. I've modeled `PQconnectdb` as `fopen`, `PQfinish` as `fclose` and call only the former in `main`. I've tried both: compiling my model with `models-mode` and copying specs to the hardcoded path and compiling the model together with `main`. In both cases `infer` didn't find the resource leak.

In the end, It would be convenient then to distribute models separately. In the infers' source, I've seen at least two options that look relevant: `specs-library` and `zip-specs-library` but infer does not recognize them or I using them wrong.

**Comment by akotulski** [comment]
Thanks for the report, we'll look into it.

Regarding writing models, I was able to produce resource leak by writing following code `PQconnectdb` (I didn't include `libpq` so it's not exactly the same). Could you try that one and see if it helps?
```
struct PGconn;
// you may have to replace 'struct PGconn' with 'PGconn'
struct PGconn *PQconnectdb(const char *conninfo) {
  return (struct PGconn*) fopen("","");
}

int main() {
  PQconnectdb("");
  return 0;
}
```
We use similar mechanism in current models: https://github.com/facebook/infer/blob/master/infer/models/c/src/libc_basic.c#L450

**Comment by akotulski** [comment]
Actually `specs-library` works as intended (`--models-mode` is not necessary I think). To reproduce it I created:
1. `pq.h` - library header, you won't need it if you are using real `libpq`
2. `pq.c` - library model implementation. This is where you put your models
3. `pq_client.c` - library user code.

Then I ran infer twice:
1. build specs in separate directory by running `infer -o pq-out -- clang -c pq.c`
2. build project and use specs from previous step: `infer --specs-library pq-out -- clang -c pq_client.c`
```
akotulski-mbp1:c akotulski$ cat pq.h 

struct PGconn;
struct PGconn *PQconnectdb(const char *conninfo);


akotulski-mbp1:c akotulski$ cat pq.c 

#include "pq.h"
struct PGconn *PQconnectdb(const char *conninfo) {
  return (struct PGconn*) fopen("","");
}

akotulski-mbp1:c akotulski$ cat pq_client.c 
#include "pq.h"

int main() {
  PQconnectdb("");
  return 0;
}

akotulski-mbp1:c akotulski$ infer -o pq_out -- clang -c pq.c 
Infer version v0.9.2-b0d2996
Copyright 2009 - present Facebook. All Rights Reserved.
Starting analysis (Infer version v0.9.2-b0d2996)

legend:
  "F" analyzing a file
  "." analyzing a procedure

F.
Analyzed 1 file

  No issues found  

akotulski-mbp1:c akotulski$ infer --specs-library pq_out/specs/ -- clang -c pq_client.c 
Infer version v0.9.2-b0d2996
Copyright 2009 - present Facebook. All Rights Reserved.
Starting analysis (Infer version v0.9.2-b0d2996)

legend:
  "F" analyzing a file
  "." analyzing a procedure

F.
Analyzed 1 file

Found 1 issue

pq_client.c:4: error: RESOURCE_LEAK
   resource acquired by call to PQconnectdb() at line 4, column 3 is not released after line 4, column 3
  2.   
  3.   int main() {
  4. >   PQconnectdb("");
  5.     return 0;
  6.   }
  7.   

Summary of the reports

  RESOURCE_LEAK: 1
```

**Comment by sumerman** [comment]
Thank you for the detailed explanation. The actual reason for `infer` not finding any issues in my case was the scope of the connection variable: it was a static global. I understand that analysis might be impossible for global variables, but expected it to work for a static global initialized with `NULL`. Should I file another issue for that or it's definitely "won't fix"?

**Comment by sumerman** [comment]
Unfortunately, even after wrapping a connection into a struct (I need a struct because there are two connections actually in my toy program) infer has ignored [at least one code path with an apparent issue](https://bitbucket.org/sumerman/pgtask/src/1308927e613c5663775262d5750088bceaf8fd9c/main.c?fileviewer=file-view-default#main.c-133).

**Comment by akotulski** [comment]
globals are very tricky to get right and we prefer to just treat them as "safe" to avoid false positives. I'd say it has low priority for us right now.

Could you share your model code as well? I understand you need to model `PQfinish` and `PQconnectdb` to get any useful reports

**Comment by sumerman** [comment]
Sure, but wouldn't it be possible to model a static global an argument that is passed to every function that is defined within a module?

[Here is my model.](https://bitbucket.org/sumerman/pgtask/src/1308927e613c5663775262d5750088bceaf8fd9c/infer_model/libpq.c?at=infer&fileviewer=file-view-default)

**Comment by akotulski** [comment]
Regarding globals - there are many options to treat global variables better, your suggestion is one of them. The problem is that tracking many variables in each function can cause performance issues and/or confuse analyzer.

I reproduced your problem and unfortunately it's known limitation of C analysis - infer doesn't treat structs in C very well. In particular, pass-by-value and return-by-value confuse our analyzer right now. In your code, `exit_nicely` confuses infer and then when you call it in line 66, analyzer gives up with `PRECONDITION_NOT_MET ` error. Minimized example looks like this:
https://gist.github.com/akotulski/7e73ef3ef8b0f9d9d1155dc1843b4dd3

Structs in C++ have much better support and the same example works for me in C++ mode, but I understand it may not be an option to you. I tested example above in C++ mode (`infer --cxx -- clang -x c++ -c test.c`) and it reported resource leak.

There is open task to address those problems with structs in C, but no one is actively working on it right now.
